### PR TITLE
Implement backend integration for tenant-aware login

### PR DESCRIPTION
**Pull request by nsklikas**
**Description**

This PR introduces the backend mechanisms necessary to support Tenant-Aware Login

#### Key Changes
- **Tenant Lookup Endpoint (`GET /api/v0/tenants/lookup`)**: 
  - Added an unauthenticated endpoint returning only the active, enabled tenants for a given email.
  - Mitigates enumeration vectors by returning an empty array for unknown emails rather than tossing out `404` errors.
- **Login Webhook  (`POST /api/v0/webhooks/login`)**:
  - Intercepts incoming Kratos sessions. When a `tenant_id` is supplied, it strictly verifies active membership.
- **Token Webhook Claims Injection (`POST /api/v0/webhooks/token`)**: 
  - Re-architected to act as the **final arbiter**. 
  - It expects the Login UI backend to establish the user's intent via Hydra's consent accept (`accept_login_request` context). It securely re-validates the membership for `tenant_id` and is solely responsible for injecting it as a claim into the final `id_token` and `access_token` session payload.
- **Development & Config Enhancements**: Updated `docker-compose.dev.yml` and Traefik routing to natively support and test the Login UI against these new hooks. The login UI was built from https://github.com/canonical/identity-platform-login-ui/compare/tenant-demo?expand=1

#### Documentation & Quality
- Added comprehensive unit tests targeting both gRPC handlers and the new HTTP edge-cases (`pkg/tenant/handlers_test.go`).
- Enriched all newly defined struct/functions with standard Canonical Godocs.
- Updated **ADR 0008** (`docs/adr/0008-tenant-aware-login.md`) highlighting the final arbiter paradigm and detailing how the selection state operates via securely encrypted frontend cookies mapped to `login_challenge`.
- Overhauled `docs/SEQUENCE_DIAGRAMS.md` with an accurate Mermaid flow detailing the 3 selection scenarios (0 tenants, 1 mapped, 2+ mapped).
- Added `docs/login-ui-prompt.md` holding the frontend's expected integration contract.

**Fixes**: https://github.com/canonical/tenant-service/issues/15

### CODE REVIEW

File: pkg/webhooks/service.go [hunk 1]
```diff
@@ -154,10 +146,122 @@ func (s *Service) HandleTokenHook(ctx context.Context, req *oauth2.TokenHookRequ
...
+		)
+		if err := s.HandleRegistration(ctx, identityID, email); err != nil {
+			s.recordError(span, "login hook: lazy reconciliation failed", err, "identity_id", identityID)
+			if counterErr := s.monitor.IncrementCounter(map[string]string{"operation": "login_reconciliation_error", "identity_id": identityID}); counterErr != nil {
+				s.logger.Errorw("failed to increment reconciliation error counter", "error", counterErr)
+			}
+			return fmt.Errorf("failed to reconcile orphaned identity: %w", err)
+		}
+		s.logger.Infow("login hook: lazy reconciliation succeeded", "identity_id", identityID)
+	}
```
Comment (nsklikas) [inline pkg/webhooks/service.go]: This is used to create a tenant for any user that has no registered tenants, perhaps it is out of place (in case something went wrong during registration). Should I remove this and instead implement it with a asynchronous reconciliation mechanism?

File: internal/storage/storage.go [hunk 2]
```diff
@@ -300,6 +300,53 @@ func (s *Storage) GetMemberByTenantAndUserID(ctx context.Context, tenantID, user
...
+	defer span.End()
+
+	var count int
+	err := s.db.Statement(ctx).
+		Select("COUNT(*)").
+		From("memberships").
+		Where(sq.Eq{"kratos_identity_id": identityID}).
+		Limit(1).
+		QueryRowContext(ctx).
+		Scan(&count)
```
Comment (shipperizer) [inline internal/storage/storage.go]: depending on the version of postgresql these calls might cripple the db, are we doing them unbound on the non-authorized endpoint?

if yes, let's maybe add a metric in here or a loud commend

File: internal/storage/storage.go [hunk 3]
```diff
@@ -300,6 +300,53 @@ func (s *Storage) GetMemberByTenantAndUserID(ctx context.Context, tenantID, user
...
+
+	if err != nil {
+		return false, fmt.Errorf("failed to check membership existence: %v", err)
+	}
+
+	return count > 0, nil
```
Comment (shipperizer) [inline internal/storage/storage.go]: if the idea is to have a bool, probably worth using the `HAVING` clause and let the db do it, at that point you should be able to drop the count

File: pkg/web/router.go [hunk 4]
```diff
@@ -69,9 +70,12 @@ func NewRouter(
...
+	// Unauthenticated tenant lookup — used by the Login UI before the user has a token.
+	// See ADR 0008 for security trade-offs. Rate limiting should be enforced at the proxy/gateway layer.
+
+	// Protected routes — lookup is excluded so the Login UI can call it without a Bearer token.
 	authRouter := chi.NewRouter()
-	authRouter.Use(authMiddleware.Authenticate())
+	authRouter.Use(authMiddleware.AuthenticateExcluding("/api/v0/tenants/lookup"))
 	authRouter.Mount("/", gRPCGatewayMux)
 
 	router.Mount("/", authRouter)
```
Comment (shipperizer) [inline pkg/web/router.go:91]: we are mounting router on top authRouter on top of grpcGatewayMux...i'm confused

File: pkg/webhooks/handlers.go [hunk 5]
```diff
@@ -73,3 +84,41 @@ func (a *API) registration(w http.ResponseWriter, r *http.Request) {
...
+			return
+		}
+		a.logger.Errorw("login hook: service error",
+			"identity_id", payload.IdentityID,
+			"tenant_id", payload.TenantID,
+			"error", err,
+		)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
```
Comment (shipperizer) [inline pkg/webhooks/handlers.go]: might be worth going for a `if err ==nil` and od the happy case, then get  back into the error case removing one nesting

File: pkg/webhooks/service.go [hunk 6]
```diff
@@ -154,10 +146,122 @@ func (s *Service) HandleTokenHook(ctx context.Context, req *oauth2.TokenHookRequ
...
+				"tenant_id", tenantID,
+			)
+			return nil, ErrNotMember
+		}
+		s.recordError(span, "token hook: failed to validate tenant membership", err,
+			"user_id", userID,
+			"tenant_id", tenantID,
+		)
+		return nil, fmt.Errorf("failed to validate tenant membership: %w", err)
+	}
```
Comment (shipperizer) [inline pkg/webhooks/service.go:189]: same comment as above, a tad bit hard to read, not a biggie as the one above though

File: pkg/webhooks/service.go (hunk 1 above)
Comment (shipperizer) [reply pkg/webhooks/service.go]: what's the potential attack vector here?can it be hijacked?

how would async look like?if we have a potential process for doing that maybe that's a safer bet

Comment (nsklikas) [reply pkg/webhooks/service.go]: As discussed, removed it

File: pkg/web/router.go (hunk 4 above)
Comment (nsklikas) [reply pkg/web/router.go:91]: Added comment

File: internal/storage/storage.go (hunk 2 above)
Comment (nsklikas) [reply internal/storage/storage.go]: Removed

File: pkg/webhooks/handlers.go (hunk 5 above)
Comment (nsklikas) [reply pkg/webhooks/handlers.go]: I tried to keep the happy path on the outer block, while reducing nesting. If you don't like the current approach, no problem with moving the happy path in the `if`

File: pkg/tenant/service.go [hunk 7]
```diff
@@ -520,3 +522,34 @@ func (s *Service) incrementCounter(operation, role string) {
 		s.logger.Warnf("failed to increment counter %s: %v", operation, err)
 	}
 }
+
+// LookupTenantsByEmail returns the enabled tenants that the given email belongs to.
+// If the email is not known to Kratos, an empty slice is returned (no error).
+// This method is called by the unauthenticated tenant lookup endpoint used by the Login UI.
+func (s *Service) LookupTenantsByEmail(ctx context.Context, email string) ([]*types.Tenant, error) {
```
Comment (BarcoMasile) [inline pkg/tenant/service.go:531]: micro-nitpick: you could technically add success/error states to the span

```go
span.SetStatus(codes.Ok, "Success")

// and with the record error you have already...
span.RecordError(err)
span.SetStatus(codes.Error, err.Error())
```

File: pkg/webhooks/handlers.go [hunk 8]
```diff
@@ -5,28 +5,47 @@ package webhooks
...
+		a.middleware = middleware
 	}
+
+	a.logger = logger
+
+	return a
 }
 
 func (a *API) RegisterEndpoints(mux *chi.Mux) {
+	if a.middleware != nil {
```
Comment (BarcoMasile) [inline pkg/webhooks/handlers.go:44]: super micro small nitpick : I would have probably kept the middleware logic outside of this, registering this API token just like the other middlerware is handled/registered

File: pkg/web/router.go (hunk 4 above)
Comment (BarcoMasile) [reply pkg/web/router.go:91]: unfortunately the gRPCGatewayMux is not a router so it needs to be "wrapper" with a chi Router to be usable with middlewares etc

**Review by BarcoMasile — DISMISSED** [review]

File: pkg/tenant/service.go (hunk 7 above)
Comment (nsklikas) [reply pkg/tenant/service.go:531]: `s.recordError` should take care of this

File: internal/storage/storage.go [hunk 9]
```diff
@@ -193,7 +210,8 @@ func (s *Storage) ListTenantsByUserID(ctx context.Context, userID string, option
 	return tenants, nextPageToken, nil
 }
 
-func (s *Storage) listTenantsByUserID(ctx context.Context, userID string, showDisabled bool) ([]*types.Tenant, error) {
+func (s *Storage) listTenantsByUserID(ctx context.Context, userID string, showDisabled bool) (tenants []*types.Tenant, err error) {
+	defer func(start time.Time) { s.observeLatency(start, "listTenantsByUserID", &err) }(time.Now())
```
Comment (wood-push-melon) [inline internal/storage/storage.go]: a minor question: should we also create a span here or it's designed to skip it for non-public methods?

File: pkg/tenant/service.go [hunk 10]
```diff
@@ -520,3 +522,34 @@ func (s *Service) incrementCounter(operation, role string) {
...
+	ctx, span := s.tracer.Start(ctx, "tenant.Service.LookupTenantsByEmail")
+	defer span.End()
+
+	s.logger.Debugw("looking up tenants by email", "email", email)
```
Comment (wood-push-melon) [inline pkg/tenant/service.go]: A minor question: should we skip logging the email address here?

File: pkg/webhooks/service.go [hunk 11]
```diff
@@ -154,10 +159,99 @@ func (s *Service) HandleTokenHook(ctx context.Context, req *oauth2.TokenHookRequ
...
 }
+
+// HandleLoginHook validates that the given identity is permitted to log in.
+//
+// If tenantID is non-empty, it checks that the identity is an active member of
+// that tenant, returning ErrNotMember if not.
+//
+// If tenantID is empty, login is allowed. This handles users who do not yet
+// have an active tenant (pending activation).
+func (s *Service) HandleLoginHook(ctx context.Context, identityID, email, tenantID string) error {
```
Comment (wood-push-melon) [inline pkg/webhooks/service.go:205]: Have a question here: do we implement the "lazy reconciliation" here or somewhere else?

I am confused by the ADR, the PR description, and the code implementation. The ADR shows that it decided not to do that. The PR description says it's implemented. And code is not implemented. I guess there might be a doc drift?

Comment (nsklikas) [reply pkg/webhooks/service.go:205]: I implemented it when I first created the PR, but then discussed it with @shipperizer and we decided that it is not the right way to solve this and I removed it. I forgot to update the PR description.

Comment (wood-push-melon) [reply pkg/webhooks/service.go:205]: oh, gotcha

**Review by wood-push-melon — DISMISSED** [review]

File: internal/storage/storage.go [hunk 12]
```diff
@@ -41,7 +42,19 @@ func NewStorage(c db.DBClientInterface, tracer tracing.TracingInterface, monitor
 	return s
 }
 
-func (s *Storage) CreateTenant(ctx context.Context, t *types.Tenant) (*types.Tenant, error) {
+func (s *Storage) observeLatency(start time.Time, operation string, err *error) {
```
Comment (BarcoMasile) [inline internal/storage/storage.go]: suggestion: `recordLatencyFor(operation string, start time.Time, err error)` (if error here is the interface error from the standard library)

File: internal/storage/storage.go [hunk 13]
```diff
@@ -300,7 +320,36 @@ func (s *Storage) GetMemberByTenantAndUserID(ctx context.Context, tenantID, user
 	return &m, nil
 }
 
-func (s *Storage) AddMember(ctx context.Context, tenantID, userID, role string) (string, error) {
+func (s *Storage) GetActiveMemberByTenantAndUserID(ctx context.Context, tenantID, userID string) (membership *types.Membership, err error) {
+	defer func(start time.Time) { s.observeLatency(start, "GetActiveMemberByTenantAndUserID", &err) }(time.Now())
```
Comment (BarcoMasile) [inline internal/storage/storage.go]: This is nice. I checked to make sure I didn't remember wrong.
This can be simplified to 
```suggestion
	defer s.observeLatency(time.Now(), "GetActiveMemberByTenantAndUserID", &err)
```
as the parameters get evaluted immediately - [link](https://go.dev/blog/defer-panic-and-recover#:~:text=A%20deferred%20function%E2%80%99s%20arguments%20are%20evaluated%20when%20the%20defer%20statement%20is%20evaluated.) to doc

**Review by BarcoMasile — APPROVED** [review]

**Review by wood-push-melon — DISMISSED** [review]

File: internal/storage/storage.go (hunk 13 above)
Comment (nsklikas) [reply internal/storage/storage.go]: changed

File: internal/storage/storage.go (hunk 12 above)
Comment (nsklikas) [reply internal/storage/storage.go]: fixed

**Review by shipperizer — APPROVED** [review]
